runtime/src/main/java/org/corfudb/runtime/collections/Query.java

```diff
@@ -0,0 +1,449 @@
+            @Nonnull final BiFunction<Val1, Val2, T> joinFunction,
+            final Function<T, U> joinProjection) {
+        return executeJoinQuery(
+                tableName1,
+                tableName2,
+                query1,
+                query2,
+                DEFAULT_OPTIONS,
+                DEFAULT_OPTIONS,
+                joinPredicate,
+                joinFunction,
+                joinProjection);
+    }
+
+    /**
+     * Execute a join of 2 tables.
+     *
+     * @param tableName1     Table name 1.
+     * @param tableName2     Table name 2.
+     * @param query1         Predicate to filter entries in table 1.
+     * @param query2         Predicate to filter entries in table 2.
+     * @param queryOptions1  Query options to transform table 1 filtered values.
+     * @param queryOptions2  Query options to transform table 2 filtered values.
+     * @param joinPredicate  Predicate to filter entries during the join.
+     * @param joinFunction   Function to merge entries.
+     * @param joinProjection Project the merged entries.
+     * @param <Val1>         Type of Value in table 1.
+     * @param <Val2>         Type of Value in table 2.
+     * @param <R>            Type of projected values from table 1 from type V.
+     * @param <S>            Type of projected values from table 2 from type W.
+     * @param <T>            Type of resultant value after merging type R and type S.
+     * @param <U>            Type of value projected from T.
+     * @return Result of query.
+     */
+    @Nonnull
+    public <Val1 extends Message, Val2 extends Message, Meta1 extends Message, Meta2 extends Message, R, S, T, U>
+    QueryResult<U> executeJoinQuery(
+            @Nonnull final String tableName1,
+            @Nonnull final String tableName2,
+            @Nonnull final Predicate<CorfuRecord<Val1, Meta1>> query1,
+            @Nonnull final Predicate<CorfuRecord<Val2, Meta2>> query2,
+            @Nonnull final QueryOptions<Val1, Meta1, R> queryOptions1,
+            @Nonnull final QueryOptions<Val2, Meta2, S> queryOptions2,
+            @Nonnull final BiPredicate<R, S> joinPredicate,
+            @Nonnull final BiFunction<R, S, T> joinFunction,
+            final Function<T, U> joinProjection) {
+
+        Collection<CorfuRecord<Val1, Meta1>> filterResult1 = scanAndFilter(tableName1, queryOptions1.getTimestamp(), query1);
+        Collection<R> queryResult1 = transform(
+                filterResult1,
+                initializeResultCollection(queryOptions1),
+                queryOptions1.getProjection());
+
+        Collection<CorfuRecord<Val2, Meta2>> filterResult2 = scanAndFilter(tableName2, queryOptions2.getTimestamp(), query2);
+        Collection<S> queryResult2 = transform(
+                filterResult2,
+                initializeResultCollection(queryOptions2),
+                queryOptions2.getProjection());
+
+        Collection<T> joinResult = new ArrayList<>();
+
+        for (R value1 : queryResult1) {
```

Comment:
As discussed, the ordering will be lost after join, right now it's okay as we don't have the requirement, later we can add a joinOption if needed. Just bookkeeping the discussion here.

Reply:
sure. thanks